akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -0,0 +1,255 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.stage
+
+import akka.stream._
+import akka.stream.impl.StreamLayout.Module
+import akka.stream.impl.fusing.{ GraphModule, GraphInterpreter }
+import akka.stream.impl.fusing.GraphInterpreter.GraphAssembly
+
+/**
+ * A GraphStage represents a reusable graph stream processing stage. A GraphStage consists of a [[Shape]] which describes
+ * its input and output ports and a factory function that creates a [[GraphStageLogic]] which implements the processing
+ * logic that ties the ports together.
+ */
+abstract class GraphStage[S <: Shape] extends Graph[S, Unit] {
+  def shape: S
+  def createLogic: GraphStageLogic
+
+  final override private[stream] lazy val module: Module = {
+    val connectionCount = shape.inlets.size + shape.outlets.size
+    val assembly = GraphAssembly(
+      Array(this),
+      Array.ofDim(connectionCount),
+      Array.fill(connectionCount)(-1),
+      Array.ofDim(connectionCount),
+      Array.fill(connectionCount)(-1))
+
+    for ((inlet, i) ← shape.inlets.iterator.zipWithIndex) {
+      assembly.ins(i) = inlet
+      assembly.inOwners(i) = 0
+    }
+
+    for ((outlet, i) ← shape.outlets.iterator.zipWithIndex) {
+      assembly.outs(i + shape.inlets.size) = outlet
+      assembly.outOwners(i + shape.inlets.size) = 0
+    }
+
+    GraphModule(assembly, shape, Attributes.none)
+  }
+
+  /**
+   * This method throws an [[UnsupportedOperationException]] by default. The subclass can override this method
+   * and provide a correct implementation that creates an exact copy of the stage with the provided new attributes.
+   */
+  override def withAttributes(attr: Attributes): Graph[S, Unit] =
+    throw new UnsupportedOperationException(
+      "withAttributes not supported by default by FlexiMerge, subclass may override and implement it")
```

Comment:
I was thinking about the delegate thing, but that would change the type of the class which can be confusing. Might be acceptable though. The current throwing approach is basically what FlexiMerge takes, and I just used it for now until we figure it out properly.